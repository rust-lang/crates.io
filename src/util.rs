pub use self::io_util::{read_fill, read_le_u32};
pub use self::request_helpers::*;
pub use crates_io_database::utils::token;

pub mod diesel;
pub mod errors;
mod io_util;
mod request_helpers;
pub mod string_excl_null;
pub mod tracing;
