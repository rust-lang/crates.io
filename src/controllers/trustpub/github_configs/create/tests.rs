use crate::tests::builders::CrateBuilder;
use crate::tests::util::{RequestHelper, Response, TestApp};
use anyhow::anyhow;
use bytes::Bytes;
use crates_io_database::models::token::{CrateScope, EndpointScope};
use crates_io_database::schema::{emails, trustpub_configs_github};
use crates_io_github::{GitHubError, GitHubUser, MockGitHubClient};
use diesel::prelude::*;
use diesel_async::RunQueryDsl;
use insta::{assert_json_snapshot, assert_snapshot};
use serde_json::json;

const URL: &str = "/api/v1/trusted_publishing/github_configs";

const CRATE_NAME: &str = "foo";

fn simple_github_mock() -> MockGitHubClient {
    let mut github_mock = MockGitHubClient::new();
    github_mock.expect_get_user().returning(|login, _| {
        Ok(GitHubUser {
            avatar_url: None,
            email: None,
            id: 42,
            login: login.into(),
            name: None,
        })
    });
    github_mock
}

async fn run_test(payload: impl Into<Bytes>) -> (TestApp, Response<()>) {
    async fn inner(payload: Bytes) -> (TestApp, Response<()>) {
        let (app, _client, cookie_client) = TestApp::full()
            .with_github(simple_github_mock())
            .with_user()
            .await;

        let mut conn = app.db_conn().await;

        CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
            .build(&mut conn)
            .await
            .unwrap();

        (app, cookie_client.post::<()>(URL, payload).await)
    }

    inner(payload.into()).await
}

#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    assert_snapshot!(app.emails_snapshot().await);

    let mut conn = app.db_conn().await;
    let config_ids = trustpub_configs_github::table
        .select(trustpub_configs_github::id)
        .get_results::<i32>(&mut conn)
        .await?;

    assert_eq!(config_ids.len(), 1);
    assert_eq!(config_ids[0], 1);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path_with_environment() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": "production",
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_empty_body() -> anyhow::Result<()> {
    let (_app, response) = run_test("").await;
    assert_snapshot!(response.status(), @"415 Unsupported Media Type");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Expected request with `Content-Type: application/json`"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_empty_json_object() -> anyhow::Result<()> {
    let (_app, response) = run_test("{}").await;
    assert_snapshot!(response.status(), @"422 Unprocessable Entity");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Failed to deserialize the JSON body into the target type: missing field `github_config` at line 1 column 2"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_owner() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "ยง$%&",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Invalid GitHub repository owner name"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_repo() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "@foo",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Invalid GitHub repository name"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_workflow() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "workflows/ci.json",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Workflow filename must end with `.yml` or `.yaml`"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_environment() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": "",
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Environment name may not be empty (use `null` to omit)"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unauthenticated() -> anyhow::Result<()> {
    let (app, client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this action requires authentication"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_legacy_token_auth() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_token()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_trusted_publishing_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from(CRATE_NAME).unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_without_trusted_publishing_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from(CRATE_NAME).unwrap()]),
            Some(vec![EndpointScope::PublishUpdate]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this token does not have the required permissions to perform this action"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_wrong_crate_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from("other-crate").unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this token does not have the required permissions to perform this action"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_wildcard_crate_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from("*").unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_missing_crate() -> anyhow::Result<()> {
    let (_app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"404 Not Found");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"crate `foo` does not exist"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_non_owner() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let other_client = app.db_new_user("other_user").await;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = other_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You are not an owner of this crate"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unknown_github_user() -> anyhow::Result<()> {
    let mut github_mock = MockGitHubClient::new();
    github_mock
        .expect_get_user()
        .returning(|login, _| Err(GitHubError::NotFound(anyhow!("User {} not found", login))));

    let (app, _client, cookie_client) = TestApp::full().with_github(github_mock).with_user().await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Unknown GitHub user or organization"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_github_error() -> anyhow::Result<()> {
    let mut github_mock = MockGitHubClient::new();
    github_mock
        .expect_get_user()
        .returning(|_, _| Err(GitHubError::Other(anyhow!("Internal Server Error"))));

    let (app, _client, cookie_client) = TestApp::full().with_github(github_mock).with_user().await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"500 Internal Server Error");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Internal Server Error"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unverified_email() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    diesel::update(emails::table.filter(emails::user_id.eq(cookie_client.as_model().id)))
        .set(emails::verified.eq(false))
        .execute(&mut conn)
        .await?;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You must verify your email address to create a Trusted Publishing config"}]}"#);

    Ok(())
}
