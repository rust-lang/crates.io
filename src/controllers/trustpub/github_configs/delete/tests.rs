use crate::tests::builders::CrateBuilder;
use crate::tests::util::{RequestHelper, TestApp};
use crates_io_database::models::Crate;
use crates_io_database::models::trustpub::{GitHubConfig, NewGitHubConfig};
use crates_io_database::schema::trustpub_configs_github;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use http::StatusCode;
use insta::assert_snapshot;
use serde_json::json;

const BASE_URL: &str = "/api/v1/trusted_publishing/github_configs";
const CRATE_NAME: &str = "foo";

fn delete_url(id: i32) -> String {
    format!("{BASE_URL}/{id}")
}

async fn create_crate(conn: &mut AsyncPgConnection, author_id: i32) -> anyhow::Result<Crate> {
    CrateBuilder::new(CRATE_NAME, author_id).build(conn).await
}

async fn create_config(conn: &mut AsyncPgConnection, crate_id: i32) -> QueryResult<GitHubConfig> {
    let config = NewGitHubConfig {
        crate_id,
        repository_owner: "rust-lang",
        repository_owner_id: 42,
        repository_name: "foo-rs",
        workflow_filename: "publish.yml",
        environment: None,
    };

    config.insert(conn).await
}

async fn get_all_configs(conn: &mut AsyncPgConnection) -> QueryResult<Vec<GitHubConfig>> {
    trustpub_configs_github::table
        .select(GitHubConfig::as_select())
        .load::<GitHubConfig>(conn)
        .await
}

/// Delete the config with a valid user that is an owner of the crate.
#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full().with_user().await;
    let mut conn = app.db_conn().await;

    let krate = create_crate(&mut conn, cookie_client.as_model().id).await?;
    let config = create_config(&mut conn, krate.id).await?;

    let response = cookie_client.delete::<()>(&delete_url(config.id)).await;
    assert_eq!(response.status(), StatusCode::NO_CONTENT);
    assert_eq!(response.text(), "");

    // Verify the config was deleted from the database
    let configs = get_all_configs(&mut conn).await?;
    assert_eq!(configs.len(), 0);

    // Verify emails were sent to crate owners
    assert_snapshot!(app.emails_snapshot().await);

    Ok(())
}

/// Try to delete the config with an unauthenticated client.
#[tokio::test(flavor = "multi_thread")]
async fn test_unauthenticated() -> anyhow::Result<()> {
    let (app, client, cookie_client) = TestApp::full().with_user().await;
    let mut conn = app.db_conn().await;

    let krate = create_crate(&mut conn, cookie_client.as_model().id).await?;
    let config = create_config(&mut conn, krate.id).await?;

    let response = client.delete::<()>(&delete_url(config.id)).await;
    assert_eq!(response.status(), StatusCode::FORBIDDEN);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this action requires authentication"}]}"#);

    // Verify the config was not deleted
    let configs = get_all_configs(&mut conn).await?;
    assert_eq!(configs.len(), 1);

    // Verify no emails were sent to crate owners
    assert_eq!(app.emails().await.len(), 0);

    Ok(())
}

/// Try to delete the config with API token authentication.
#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full().with_token().await;
    let mut conn = app.db_conn().await;

    let krate = create_crate(&mut conn, cookie_client.as_model().id).await?;
    let config = create_config(&mut conn, krate.id).await?;

    let response = token_client.delete::<()>(&delete_url(config.id)).await;
    assert_eq!(response.status(), StatusCode::FORBIDDEN);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this action can only be performed on the crates.io website"}]}"#);

    // Verify the config was not deleted
    let configs = get_all_configs(&mut conn).await?;
    assert_eq!(configs.len(), 1);

    // Verify no emails were sent to crate owners
    assert_eq!(app.emails().await.len(), 0);

    Ok(())
}

/// Try to delete a config that does not exist.
#[tokio::test(flavor = "multi_thread")]
async fn test_config_not_found() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full().with_user().await;

    let response = cookie_client.delete::<()>(&delete_url(42)).await;
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Not Found"}]}"#);

    // Verify no emails were sent to crate owners
    assert_eq!(app.emails().await.len(), 0);

    Ok(())
}

/// Try to delete the config with a user who is not an owner of the crate.
#[tokio::test(flavor = "multi_thread")]
async fn test_non_owner() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full().with_user().await;
    let mut conn = app.db_conn().await;

    let krate = create_crate(&mut conn, cookie_client.as_model().id).await?;
    let config = create_config(&mut conn, krate.id).await?;

    // Create another user who is not an owner of the crate
    let other_client = app.db_new_user("other_user").await;

    let response = other_client.delete::<()>(&delete_url(config.id)).await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You are not an owner of this crate"}]}"#);

    // Verify the config was not deleted
    let configs = get_all_configs(&mut conn).await?;
    assert_eq!(configs.len(), 1);

    // Verify no emails were sent to crate owners
    assert_eq!(app.emails().await.len(), 0);

    Ok(())
}

/// Try to delete the config with a user that is part of a team that owns
/// the crate.
#[tokio::test(flavor = "multi_thread")]
async fn test_team_owner() -> anyhow::Result<()> {
    let (app, _client) = TestApp::full().empty().await;
    let mut conn = app.db_conn().await;

    let user = app.db_new_user("user-org-owner").await;
    let user2 = app.db_new_user("user-one-team").await;

    let krate = create_crate(&mut conn, user.as_model().id).await?;
    let config = create_config(&mut conn, krate.id).await?;

    let body = json!({ "owners": ["github:test-org:all"] }).to_string();
    let response = user.put::<()>("/api/v1/crates/foo/owners", body).await;
    assert_eq!(response.status(), StatusCode::OK);

    let response = user2.delete::<()>(&delete_url(config.id)).await;
    assert_eq!(response.status(), StatusCode::BAD_REQUEST);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You are not an owner of this crate"}]}"#);

    // Verify the config was not deleted
    let configs = get_all_configs(&mut conn).await?;
    assert_eq!(configs.len(), 1);

    // Verify no emails were sent to crate owners
    assert_eq!(app.emails().await.len(), 0);

    Ok(())
}
