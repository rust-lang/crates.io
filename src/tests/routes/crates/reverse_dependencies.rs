use crate::tests::builders::{CrateBuilder, VersionBuilder};
use crate::tests::util::{RequestHelper, TestApp};
use http::StatusCode;
use insta::{assert_json_snapshot, assert_snapshot};

#[tokio::test(flavor = "multi_thread")]
async fn reverse_dependencies() {
    let (app, anon, user) = TestApp::init().with_user();
    let mut conn = app.db_conn();
    let user = user.as_model();

    let c1 = CrateBuilder::new("c1", user.id).expect_build(&mut conn);

    CrateBuilder::new("c2", user.id)
        .version(VersionBuilder::new("1.0.0").dependency(&c1, None))
        .version(
            VersionBuilder::new("1.1.0")
                .dependency(&c1, None)
                .dependency(&c1, Some("foo")),
        )
        .expect_build(&mut conn);

    let response = anon
        .get::<()>("/api/v1/crates/c1/reverse_dependencies")
        .await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".versions[].created_at" => "[datetime]",
        ".versions[].updated_at" => "[datetime]",
    });

    // c1 has no dependent crates.
    let response = anon
        .get::<()>("/api/v1/crates/c2/reverse_dependencies")
        .await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".versions[].created_at" => "[datetime]",
        ".versions[].updated_at" => "[datetime]",
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn reverse_dependencies_when_old_version_doesnt_depend_but_new_does() {
    let (app, anon, user) = TestApp::init().with_user();
    let mut conn = app.db_conn();
    let user = user.as_model();

    let c1 = CrateBuilder::new("c1", user.id)
        .version("1.1.0")
        .expect_build(&mut conn);

    CrateBuilder::new("c2", user.id)
        .version("1.0.0")
        .version(VersionBuilder::new("2.0.0").dependency(&c1, None))
        .expect_build(&mut conn);

    let response = anon
        .get::<()>("/api/v1/crates/c1/reverse_dependencies")
        .await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".versions[].created_at" => "[datetime]",
        ".versions[].updated_at" => "[datetime]",
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn reverse_dependencies_when_old_version_depended_but_new_doesnt() {
    let (app, anon, user) = TestApp::init().with_user();
    let mut conn = app.db_conn();
    let user = user.as_model();

    let c1 = CrateBuilder::new("c1", user.id)
        .version("1.0.0")
        .expect_build(&mut conn);

    CrateBuilder::new("c2", user.id)
        .version(VersionBuilder::new("1.0.0").dependency(&c1, None))
        .version("2.0.0")
        .expect_build(&mut conn);

    let response = anon
        .get::<()>("/api/v1/crates/c1/reverse_dependencies")
        .await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".versions[].created_at" => "[datetime]",
        ".versions[].updated_at" => "[datetime]",
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn prerelease_versions_not_included_in_reverse_dependencies() {
    let (app, anon, user) = TestApp::init().with_user();
    let mut conn = app.db_conn();
    let user = user.as_model();

    let c1 = CrateBuilder::new("c1", user.id)
        .version("1.0.0")
        .expect_build(&mut conn);

    CrateBuilder::new("c2", user.id)
        .version("1.1.0-pre")
        .expect_build(&mut conn);

    CrateBuilder::new("c3", user.id)
        .version(VersionBuilder::new("1.0.0").dependency(&c1, None))
        .version("1.1.0-pre")
        .expect_build(&mut conn);

    let response = anon
        .get::<()>("/api/v1/crates/c1/reverse_dependencies")
        .await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".versions[].created_at" => "[datetime]",
        ".versions[].updated_at" => "[datetime]",
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn yanked_versions_not_included_in_reverse_dependencies() {
    let (app, anon, user) = TestApp::init().with_user();
    let mut conn = app.db_conn();
    let user = user.as_model();

    let c1 = CrateBuilder::new("c1", user.id)
        .version("1.0.0")
        .expect_build(&mut conn);

    CrateBuilder::new("c2", user.id)
        .version("1.0.0")
        .version(VersionBuilder::new("2.0.0").dependency(&c1, None))
        .expect_build(&mut conn);

    let response = anon
        .get::<()>("/api/v1/crates/c1/reverse_dependencies")
        .await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".versions[].created_at" => "[datetime]",
        ".versions[].updated_at" => "[datetime]",
    });

    use crate::schema::versions;
    use diesel::{ExpressionMethods, QueryDsl, RunQueryDsl};

    diesel::update(versions::table.filter(versions::num.eq("2.0.0")))
        .set(versions::yanked.eq(true))
        .execute(&mut conn)
        .unwrap();

    let response = anon
        .get::<()>("/api/v1/crates/c1/reverse_dependencies")
        .await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".versions[].created_at" => "[datetime]",
        ".versions[].updated_at" => "[datetime]",
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn reverse_dependencies_includes_published_by_user_when_present() {
    let (app, anon, user) = TestApp::init().with_user();
    let mut conn = app.db_conn();
    let user = user.as_model();

    use crate::schema::versions;
    use diesel::{update, ExpressionMethods, RunQueryDsl};

    let c1 = CrateBuilder::new("c1", user.id)
        .version("1.0.0")
        .expect_build(&mut conn);
    CrateBuilder::new("c2", user.id)
        .version(VersionBuilder::new("2.0.0").dependency(&c1, None))
        .expect_build(&mut conn);

    // Make c2's version (and,incidentally, c1's, but that doesn't matter) mimic a version
    // published before we started recording who published versions
    let none: Option<i32> = None;
    update(versions::table)
        .set(versions::published_by.eq(none))
        .execute(&mut conn)
        .unwrap();

    // c3's version will have the published by info recorded
    CrateBuilder::new("c3", user.id)
        .version(VersionBuilder::new("3.0.0").dependency(&c1, None))
        .expect_build(&mut conn);

    let response = anon
        .get::<()>("/api/v1/crates/c1/reverse_dependencies")
        .await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".versions[].created_at" => "[datetime]",
        ".versions[].updated_at" => "[datetime]",
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn reverse_dependencies_query_supports_u64_version_number_parts() {
    let (app, anon, user) = TestApp::init().with_user();
    let mut conn = app.db_conn();
    let user = user.as_model();

    let large_but_valid_version_number = format!("1.0.{}", u64::MAX);

    let c1 = CrateBuilder::new("c1", user.id).expect_build(&mut conn);

    // The crate that depends on c1...
    CrateBuilder::new("c2", user.id)
        // ...has a patch version at the limits of what the semver crate supports
        .version(VersionBuilder::new(&large_but_valid_version_number).dependency(&c1, None))
        .expect_build(&mut conn);

    let response = anon
        .get::<()>("/api/v1/crates/c1/reverse_dependencies")
        .await;
    assert_eq!(response.status(), StatusCode::OK);
    assert_json_snapshot!(response.json(), {
        ".versions[].created_at" => "[datetime]",
        ".versions[].updated_at" => "[datetime]",
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unknown_crate() {
    let (_, anon) = TestApp::init().empty();

    let response = anon
        .get::<()>("/api/v1/crates/unknown/reverse_dependencies")
        .await;
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"crate `unknown` does not exist"}]}"#);
}
