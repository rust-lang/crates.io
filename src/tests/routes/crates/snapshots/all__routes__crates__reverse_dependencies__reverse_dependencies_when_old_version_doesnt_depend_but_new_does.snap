---
source: src/tests/routes/crates/reverse_dependencies.rs
expression: response.json()
---
{
  "dependencies": [
    {
      "crate_id": "c1",
      "default_features": false,
      "downloads": 0,
      "features": [],
      "id": 1,
      "kind": "normal",
      "optional": false,
      "req": ">= 0",
      "target": null,
      "version_id": 3
    }
  ],
  "meta": {
    "total": 1
  },
  "versions": [
    {
      "audit_actions": [],
      "bin_names": null,
      "checksum": "                                                                ",
      "crate": "c2",
      "crate_size": 0,
      "created_at": "[datetime]",
      "dl_path": "/api/v1/crates/c2/2.0.0/download",
      "downloads": 0,
      "features": {},
      "has_lib": null,
      "id": 3,
      "lib_links": null,
      "license": null,
      "links": {
        "authors": "/api/v1/crates/c2/2.0.0/authors",
        "dependencies": "/api/v1/crates/c2/2.0.0/dependencies",
        "version_downloads": "/api/v1/crates/c2/2.0.0/downloads"
      },
      "num": "2.0.0",
      "published_by": {
        "avatar": null,
        "id": 1,
        "login": "foo",
        "name": null,
        "url": "https://github.com/foo"
      },
      "readme_path": "/api/v1/crates/c2/2.0.0/readme",
      "rust_version": null,
      "updated_at": "[datetime]",
      "yank_message": null,
      "yanked": false
    }
  ]
}
